From 679b61f0595f70758423abc65fa5f84daa615bb4 Mon Sep 17 00:00:00 2001
From: Tadeus Prastowo <eus@member.fsf.org>
Date: Tue, 14 Jun 2011 15:34:20 +0200
Subject: [PATCH] Fix the problem of outdated rq clock and wrong SCHED_HRTICK firing
 time.

---
 kernel/sched_dl.c |   20 ++++++++++++++++----
 1 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/kernel/sched_dl.c b/kernel/sched_dl.c
index 7a4c5a6..af59ac9 100644
--- a/kernel/sched_dl.c
+++ b/kernel/sched_dl.c
@@ -18,6 +18,10 @@ static long long nsec_high(unsigned long long nsec);
 static unsigned long nsec_low(unsigned long long nsec);
 #define SPLIT_NS(x) nsec_high(x), nsec_low(x)
 
+#ifdef CONFIG_SCHED_HRTICK
+#define SCHED_HRTICK_SMALLEST 10000
+#endif
+
 static inline int dl_time_before(u64 a, u64 b)
 {
 	return (s64)(a - b) < 0;
@@ -203,6 +207,13 @@ static void update_dl_entity(struct sched_dl_entity *dl_se)
 	}
 
 	if (dl_time_before(dl_se->deadline, rq->clock) ||
+#ifdef CONFIG_SCHED_HRTICK
+	    /* Even if dl_entity_overflow is false, function
+	     * start_hrtick_dl will not run, and therefore, bandwidth
+	     * enforcement failure will occur
+	     */
+	    dl_se->runtime <= SCHED_HRTICK_SMALLEST ||
+#endif
 	    dl_entity_overflow(dl_se, rq->clock)) {
 		dl_se->deadline = rq->clock + dl_se->dl_deadline;
 		dl_se->runtime = dl_se->dl_runtime;
@@ -534,10 +545,8 @@ static void check_preempt_curr_dl(struct rq *rq, struct task_struct *p,
 #ifdef CONFIG_SCHED_HRTICK
 static void start_hrtick_dl(struct rq *rq, struct task_struct *p)
 {
-	s64 delta = p->dl.dl_runtime - p->dl.runtime;
-
-	if (delta > 10000)
-		hrtick_start(rq, delta);
+	if (p->dl.runtime > SCHED_HRTICK_SMALLEST)
+		hrtick_start(rq, p->dl.runtime);
 }
 #else
 static void start_hrtick_dl(struct rq *rq, struct task_struct *p)
@@ -571,7 +580,10 @@ struct task_struct *pick_next_task_dl(struct rq *rq)
 	BUG_ON(!dl_se);
 
 	p = dl_task_of(dl_se);
+
+	update_rq_clock(rq);
 	p->se.exec_start = rq->clock;
+
 #ifdef CONFIG_SCHED_HRTICK
 	if (hrtick_enabled(rq))
 		start_hrtick_dl(rq, p);
-- 
1.7.1

